waitport.1
.TH waitport 1
.SH NAME
waitport \- wait for a local TCP service to show up
.SH SYNOPSIS
.B waitport port command ...

.B waitport ip/port command ...

.SH DESCRIPTION
.B waitport
will repeatedly do what netstat does (go through /proc/net/tcp and
/proc/net/tcp6) and look for a specific TCP port in LISTEN state to show
up. When that happens, it will run the rest of the command line.

There is a timeout after 1000 tries (10 tries per second). If the port
does not show up by then, waitport will abort with a non-zero exit code.

This program is meant for use with minit, when one service starts a
service, and another program should only be started once the first
service has bound to the port (i.e. is accepting connections).

Use waitport to resolve race conditions in dependency chains in minit.

.SH AUTHOR
waitport was written by Felix von Leitner and can be downloaded from
.I http://www.fefe.de/minit/



CHANGES
0.11:
  when serdo is run without arguments, and a file called "script"
    exists in the current working directory, executes that (to save a
    few inodes for the "params" files
  fix (apparently unexploitable) stack clobber in serdo
    (found by Nikola Vladov, many thanks)
  add ulimit to serdo
  a service can now contain a "root" symlink to where it should be
    started
  only install SIGWINCH and SIGINT if running with PID 1
  you can now do "make MINITROOT=/var/minit"
  add ftrigger (run a command if a file changes)
  add waitinterface and waitport (to resolve races)
  add powersave (demo tool to set up power saving, needs customization, not actually installed)
  add governor (demo tool to set CPU frequency governor based on AC state,
    not actually needed anymore with current hardware, not actually installed)

0.10:
  add sample script for /etc/minit/ctrlaltdel/run as
    control/ctrlaltdel-run (Florian Westphal)
  add killall5, needs /proc on Linux though (Florian Westphal)
  add shutdown (Bernd Wachter)
  add -Wall and -W and kill resulting compiler warnings
  several fixes (Bernd Wachter)
  add update support - tell minit to execve something else, a newer
    version of itself, preferably (Florian Westphal)
  also add minit-update program to download state from old minit and
    upload state to new minit after update (Florian Westphal)
  #include minit.h and use MINITROOT consistently (Lordy)
  Don't assume MINITROOT is 10 bytes long (Erich Schubert)
  add man pages!! (Erich Schubert)
  add a way to ask minit for dependencies
  msvc -H will now dump the ten least recently spawned processes.
    This is useful to see which process is looping.
  fix bug in msvc killing process group if readpid failed
    (Erich Schubert)
  You can now blacklist services temporarily by passing -servicename on
    the kernel command line.  The first time they are to be started will
    then be skipped.  Use this, for example, to not start the dhcp
    client when your notebook is not plugged in a network.
  add serdo (execute batch, no shell features whatsoever)
  make "msvc servicename" output easier to parse if not outputting to a tty
  require libowfat instead of shipping half of it ourselves
  replace malloc with alloca; minit is now 6644 bytes on i386.
  add usage for serdo (Nikola Vladov)
  fix msvc -d followed by msvc -h (would send SIGHUP to pid 1, i.e. kill
    minit; found by Paul Fox)
  added small special-purpose malloc using a static buffer before
    calling the real malloc (Nikola Vladov)

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  For stopped services, minit did not record the time when it finished,
    but when it was started.
  add hard-reboot from Tommi Virtanen.  He also asked me to include an
    explicit license statement and sent a patch to make debian packaging
    easier (deb files can't include FIFOs).
  Olaf: add write_proc.
  Olaf: lines in depends can now be commented out with a #
  Olaf: add a method to clear dead services (mark as terminated)
  minit now checks whether a PID actually exists before accepting msvc -P
    this removes a race condition when the forked service terminates
    before pidfilehack notifies minit of the PID
  msvc now accepts more than one service after -o, -d, ...
  msvc now accepts /etc/minit/sshd instead of sshd as service name
  better error handling in msvc (more and better error messages)
  minit now has a heuristic to detect time zone changes and still
    have msvc report the correct elapsed time since process start/end
    (don't look at the source code, it's ultra-kludgey)
  add msvc -g (get PID for shell scripts)

0.8:
  call waitpid repeatedly until it returns "no children".
  This reaps zombies faster.

0.7:
  found and fixed the bug that made the logging support not work when
    minit was running as PID 1.
  saw that the shutdown from sysvinit actually does what it's supposed
    to do, even with minit, if one uses the -n mode.  It should be quite
    easy to write a nice and clean shutdown for minit now.
  if a service depends on a service that already finished, don't start
    it again.

0.6.1:
  fixed msvc and minit handling of msvc -P.
  pidfilehack has now actually been tested and works with ssh.

0.6:
  add tty and session leadership magic and I actually got getty to work
    as init on my laptop, i.e. in production use!
  add msvc man page.

0.5.1:
  cut&paste error, set wrong variable to 0 for keyboard request.
    Thanks, Schulti.

0.5:
  add "log" subdirectory (see README).
  get rid of stdio and *printf and switch to routines from libowfat.

0.4:
  add pidfilehack.  It will take the first argument as filename for the
  PID file (i.e. /var/run/sshd.pid), the second argument as service name
  for minit (i.e. ssh) and the rest of the command line is run without
  path searching in a child process.  The parent process will then
  try to open the pidfile several times (for up to 30 seconds), read the
  PID off it and run msvc to tell minit the PID of the service.  minit
  can then track sshd and restart it if necessary.  This gross hack does
  not work for daemons that do not write a PID file, but there are not
  many of those, fortunately.


waitport.c
#include <unistd.h>
#include <errmsg.h>
#include <time.h>
#include <string.h>
#include <fmt.h>
#include <scan.h>
#include <ip4.h>
#include <ip6.h>
#include <stralloc.h>
#include <buffer.h>
#include <fcntl.h>
#include <ctype.h>

static int netstat(const char* addr,unsigned int wantedport) {
  /* see linux/Documentation/networking/proc_net_tcp.txt */
  int fd=-1;
  char rbuf[4096];	/* since these files can become arbitrarily large, we'll use a real buffer to read from them */
  const char* filenames[] = { "/proc/net/tcp6", "/proc/net/tcp" };
  buffer b;
  unsigned int fn;
  stralloc line;

  for (fn=0; fn<sizeof(filenames)/sizeof(filenames[0]); ++fn) {
    const char* filename=filenames[fn];

    fd=open(filename,O_RDONLY|O_CLOEXEC);
    if (fd==-1)
      continue;
    buffer_init(&b,read,fd,rbuf,sizeof(rbuf));	/* can't fail */
    for (;;) {
      int r;
      char* c;
      char* local;
      int v6;
      stralloc_zero(&line);
      if ((r=buffer_getline_sa(&b,&line))==-1) {
	close(fd);
	die(1,"read error from ",filename);
      }
      if (r==0) break;
      if (line.len < 1 || line.s[line.len-1]!='\n') {
parseerror:
	close(fd);
	die(1,"parse error in ",filename);
      }
      line.s[line.len-1]=0;	/* string is now null terminated */

      /* First token is something like "917:", skip */
      for (c=line.s; *c && *c!=':'; ++c) ;
      if (*c != ':') continue;	/* first line is boilerplate text and has no :-token, skip it */
      ++c;
      for (; *c==' '; ++c) ;
      /* Next token is something like "00000000:1770" or "00000000000000000000000000000000:0016" */
      local=c;
      for (; isxdigit(*c); ++c) ;
      if (c-local != 8 && c-local != 32)	/* we only support ipv4 and ipv6; this is neither */
	continue;
      v6=(c-local==32);
      if (*c!=':') goto parseerror;
      for (r=1; r<5; ++r) {
	if (!isxdigit(c[r])) goto parseerror;
      }
      if (c[5]!=' ') goto parseerror;
      c[5]=0;
      c+=6;
      /* Next token is the same thing, but we don't really need it, so
       * just skip till next whitespace */
      for (; *c && *c!=' '; ++c) ;
      if (*c!=' ') goto parseerror; ++c;
      /* Next is the state; if we are looking at tcp, 0A means LISTEN */
      if (filename[10]=='t' && c[0]=='0' && c[1]=='A') {
	/* TCP LISTEN */
	size_t n;
	union {
	  char ip[16];
	  uint32_t ints[4];
	} omgwtfbbq;
	char ipstring[FMT_IP6];
	unsigned short port;
	unsigned long temp;

	/* we can only be here if the hex string is 8 or 32 bytes long, see above */
	for (n=0; n<(unsigned int)v6*3+1; ++n) {
	  scan_xlongn(local+n*8,8,&temp);
	  omgwtfbbq.ints[n]=temp;
	}

	if (scan_xshort(local+(((unsigned int)v6*3+1)*8)+1,&port)!=4)	/* can't happen, we validated with isxdigit */
	  goto parseerror;

	ipstring[v6 ? fmt_ip6c(ipstring,omgwtfbbq.ip) : fmt_ip4(ipstring,omgwtfbbq.ip)]=0;

	if (!strcmp(ipstring,addr?addr : (v6?"::":"0.0.0.0")) && port==wantedport) {
	  close(fd);
	  return 1;
	}

      }
    }
    close(fd);
    fd=-1;
  }

  return 0;
}

int main(int argc,char* argv[],char* envp[]) {
  unsigned short port;

  unsigned int i;
  struct timespec req,rem;

  char* s=argv[1];
  char* t=strchr(s,'/');

  errmsg_iam("waitport");
  if (argc<2)
usage:
    die(0,"usage: waitport ::/111 some.rpcd\n\twaits for a service to bind to TCP port 111 on ::, then executes the rest of the command line");

  {
    if (t) {
      *t=0;
      if (scan_ushort(t+1,&port)==0) goto usage;
    } else {
      if (scan_ushort(s,&port)==0) goto usage;
      s=0;
    }
  }

  req.tv_sec=0; req.tv_nsec=100000000;
  for (i=0; i<1000; ++i) {
    if (netstat(s,port)) {
      if (argv[2]) {
	execve(argv[2],argv+2,envp);
	diesys(1,"execve");
      }
      return 0;
    }
    nanosleep(&req,&rem);
  }
  die(1,"service on port not showing up");
}
